### PR TITLE
[ARCTIC-1339][AMS]: AMS provider Iceberg rest catalog service

### LINKED ISSUE #1339: [Subtask]: Make AMS as an Iceberg Catalog provider via implement RestCatalogAPI
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### Description

Provider a iceberg catalog implement `com.netease.arctic.iceberg.AMSCatalog`  to visit Iceberg format table managed by AMS internal catalog

### Parent issue

https://github.com/NetEase/arctic/issues/1061

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by baiyangtx**
## Why are the changes needed?

To fix #1339 
 

## Brief change log

- Add IcebergRestCatalogService to handle iceberg rest catalog http request.
- Init http server in ArcticManagementContainer， using independent http server objecter 
- Add InternalTableOperations to support load iceberg in AMS
- Make getCatalog(name) thrift call return http-port to support RestCatalog via CatalogLoader.

## How was this patch tested?
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/catalog/InternalIcebergCatalogImpl.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import com.netease.arctic.server.persistence.mapper.TableMetaMapper;
+import com.netease.arctic.server.table.TableMetadata;
+import com.netease.arctic.server.utils.IcebergTableUtils;
+import com.netease.arctic.table.ArcticTable;
+import org.apache.iceberg.BaseTable;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.io.FileIO;
+
+public class InternalIcebergCatalogImpl extends MixedCatalogImpl {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/catalog/InternalIcebergCatalogImpl.java:17]: Why should the `InternalIcebergCatalogImpl` extend `MixedCatalogImpl`?

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+package com.netease.arctic.server.dashboard.response;
+
+
+/**
+ * iceberg rest-catalog api error response code
+ */
+public enum IcebergRestErrorCode {
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java:25]: Seems unused class.

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+import static io.javalin.apibuilder.ApiBuilder.get;
+import static io.javalin.apibuilder.ApiBuilder.head;
+import static io.javalin.apibuilder.ApiBuilder.path;
+import static io.javalin.apibuilder.ApiBuilder.post;
+
+public class IcebergRestCatalogService extends PersistentBase {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergRestCatalogService.class);
+
+  public static final String REST_CATALOG_API_PREFIX = "/api/iceberg/rest/catalog";
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java:73]: set to /api/iceberg/rest

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+
+  private final TableService tableService;
+
+  private final Set<String> catalogPropertiesNotReturned = Collections.unmodifiableSet(
+      Sets.newHashSet(CatalogMetaProperties.TABLE_FORMATS)
+  );
+
+  private final Set<String> catalogPropertiesOverwrite = Collections.unmodifiableSet(
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java:81]: set to static

File: ams/server/src/main/java/com/netease/arctic/server/catalog/InternalIcebergCatalogImpl.java (hunk 1 above)
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/catalog/InternalIcebergCatalogImpl.java:17]: to reuse the logic of  listDatabase,   exist(database)   exist(table)

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+public class IcebergRestCatalogService extends PersistentBase {
+
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergRestCatalogService.class);
+
+  public static final String REST_CATALOG_API_PREFIX = "/api/iceberg/rest/catalog";
+
+  private final TableService tableService;
+
+  private final Set<String> catalogPropertiesNotReturned = Collections.unmodifiableSet(
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java]: this properties is useless for iceberg RestCatalog

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+    BadRequest(400),
+    NotAuthorized(401),
+    Forbidden(403),
+    UnsupportedOperation(406),
+    AuthenticationTimeout(419),
+    NotFound(404),
+    Conflict(409),
+    InternalServerError(500),
+    ServiceUnavailable(503),
+
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java]: blank lines deleted

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java (hunk 2 above)
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java:25]: Used in `handleException` in this class

**Comment by baiyangtx** [comment]
AMS Table show table information
image

Visiting from terminal
image


Showing table details

image


Optimizing is ok, but with some problems in the frontend shown.

https://github.com/NetEase/arctic/pull/1633

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java (hunk 2 above)
Comment (Aireed) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java:25]: should we use code provided by javalin??

**Comment by baiyangtx** [comment]
image

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+          "multi-level namespace is not supported now");
+      catalog.createDatabase(ns.levels()[0]);
+      return CreateNamespaceResponse.builder().withNamespace(ns).build();
+    });
+  }
+
+  /**
+   * GET /{catalog}/v1/namespaces/{namespaces}
+   */
+  public void getNamespace(Context ctx) {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java:201]: I have supported getNamespace.  return an empty map.

File: ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java (hunk 2 above)
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java:25]: Those code are defined by iceberg rest api and correlation with iceberg exceptions. so i redefined the http code here.

Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/dashboard/response/IcebergRestErrorCode.java]: Remove the useless code.

File: ams/server/src/main/java/com/netease/arctic/server/catalog/InternalIcebergCatalogImpl.java (hunk 1 above)
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/catalog/InternalIcebergCatalogImpl.java]: Extract the reuse able method to `InternalCatalog` and make the `InternalIcebergCatalogImpl` to extend `InternalCatalog`

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+  }
+
+
+  /**
+   * POST PREFIX/{catalog}/v1/namespaces/{namespace}/tables/{table}/metrics
+   */
+  public void metricReport(Context ctx) {
+    handleTable(ctx, (catalog, tableMeta) -> {
+      ReportMetricsRequest request = ctx.bodyAsClass(ReportMetricsRequest.class);
+      LOG.debug(request.toString());
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java]: Add a new Logger

File: ams/server/src/main/java/com/netease/arctic/server/utils/IcebergTableUtils.java [hunk 9]
```diff
@@ -19,27 +19,58 @@
...
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
+import java.util.UUID;
 
 public class IcebergTableUtils {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/utils/IcebergTableUtils.java]: refactor the codes

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+          .build();
+    });
+  }
+
+  /**
+   * POST PREFIX/{catalog}/v1/namespaces/{namespace}/tables
+   */
+  public void createTable(Context ctx) {
+    handleNamespace(ctx, (catalog, database) -> {
+      Preconditions.checkArgument(catalog.exist(database));
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java]: add new exception and message.

File: ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/IcebergRestCatalogService.java]: change the NotReturned set to empty set.

File: ams/server/src/main/java/com/netease/arctic/server/terminal/TerminalManager.java [hunk 11]
```diff
@@ -326,6 +332,22 @@ private TerminalSessionFactory loadTerminalSessionFactory(Configurations conf) {
     return factory;
   }
 
+  private void applyClientProperties(CatalogMeta catalogMeta) {
+    Set<TableFormat> formats = CatalogUtil.tableFormats(catalogMeta);
+    if (CatalogMetaProperties.CATALOG_TYPE_AMS.equalsIgnoreCase(catalogMeta.getCatalogType())) {
+      if (formats.contains(TableFormat.ICEBERG)) {
+        int httpPort = Integer.parseInt(catalogMeta.getCatalogProperties().get(CatalogMetaProperties.HTTP_PORT));
+        catalogMeta.putToCatalogProperties(CatalogProperties.CATALOG_IMPL, RESTCatalog.class.getName());
+        catalogMeta.putToCatalogProperties("uri", "http://127.0.0.1:" + httpPort + "/api/iceberg/rest");
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/terminal/TerminalManager.java]: yes, I have replace the property

**Review by zhoujinsong — APPROVED** [review]
LGTM. Thanks a lot for your work!

**Pull request merged**